src/screentovnc.cpp

```diff
@@ -604,6 +613,181 @@ void ScreenToVnc::updateClientCursors(rfbScreenInfoPtr rfbScreen, bool emptyMous
+  for (int i=0; i < 256; i++) {
+    ioctl(uinputKeyboardDeviceFD, UI_SET_KEYBIT, i);
+  }
+
+
+  memset(&uud, 0, sizeof(uud));
+  snprintf(uud.name, UINPUT_MAX_NAME_SIZE, "uinput old interface");
+  write(uinputKeyboardDeviceFD, &uud, sizeof(uud));
+
+  ioctl(uinputKeyboardDeviceFD, UI_DEV_CREATE);
+}
+
+void ScreenToVnc::emitKeystroke(int type, int code, int val)
+{
+   struct input_event ie;
+
+   ie.type = type;
+   ie.code = code;
+   ie.value = val;
+   /* timestamp values below are ignored */
+   ie.time.tv_sec = 0;
+   ie.time.tv_usec = 0;
+
+   write(uinputKeyboardDeviceFD, &ie, sizeof(ie));
+}
+
+void ScreenToVnc::keyboardHandler(rfbBool isKeyDown, rfbKeySym keySym, rfbClientPtr cl)
+{
+    int keyCode = -1;
+    int shift = 0;
+    switch(keySym){
+      case XK_Return:          keyCode=KEY_ENTER;         shift=0;  break;
+      case XK_Escape:          keyCode=KEY_ESC;           shift=0;  break;
+      case XK_Tab:             keyCode=KEY_TAB;           shift=0;  break;
+      case XK_Caps_Lock:       keyCode=KEY_CAPSLOCK;      shift=0;  break;
+      case XK_F1:              keyCode=KEY_F1;            shift=0;  break;
+      case XK_F2:              keyCode=KEY_F2;            shift=0;  break;
+      case XK_F3:              keyCode=KEY_F3;            shift=0;  break;
+      case XK_F4:              keyCode=KEY_F4;            shift=0;  break;
+      case XK_F5:              keyCode=KEY_F5;            shift=0;  break;
+      case XK_F6:              keyCode=KEY_F6;            shift=0;  break;
+      case XK_F7:              keyCode=KEY_F7;            shift=0;  break;
+      case XK_F8:              keyCode=KEY_F8;            shift=0;  break;
+      case XK_F9:              keyCode=KEY_F9;            shift=0;  break;
+      case XK_F10:             keyCode=KEY_F10;           shift=0;  break;
+      case XK_F11:             keyCode=KEY_F11;           shift=0;  break;
+      case XK_F12:             keyCode=KEY_F12;           shift=0;  break;
+      case XK_Page_Up:         keyCode=KEY_PAGEUP;        shift=0;  break;
+      case XK_Page_Down:       keyCode=KEY_PAGEDOWN;      shift=0;  break;
+      case XK_Home:            keyCode=KEY_HOME;          shift=0;  break;
+      case XK_End:             keyCode=KEY_END;           shift=0;  break;
+      case XK_Insert:          keyCode=KEY_INSERT;        shift=0;  break;
+      case XK_Delete:          keyCode=KEY_DELETE;        shift=0;  break;
+      case XK_Left:            keyCode=KEY_LEFT;          shift=0;  break;
+      case XK_Right:           keyCode=KEY_RIGHT;         shift=0;  break;
+      case XK_Up:              keyCode=KEY_UP;            shift=0;  break;
+      case XK_Down:            keyCode=KEY_DOWN;          shift=0;  break;
+      case XK_BackSpace:       keyCode=KEY_BACKSPACE;     shift=0;  break;
+      case XK_space:           keyCode=KEY_SPACE;         shift=0;  break;
+      case XK_exclam:          keyCode=KEY_1;             shift=1;  break;
```

Comment:
This shift handling seems to be wired to only one keyboard layout as special characters may also have shift pressed, like on this line. I'm not sure we should make assumptions about selected hardware keyboard layout but it might be a reasonable assumption that it matches on client and device side (assuming that we can do it like that).

Reply:
yes, this is keyboard-layout-specific. i have only the one layout to test, but the code should be easy to adapt.

Reply:
Well, I use Finnish keyboard and I can already see that this doesn't support all of the letters, but this can be expanded and it works surprisingly well on non-qwerty layout too.